app_dart/lib/src/request_handlers/get_status_firestore.dart

```diff
@@ -0,0 +1,97 @@
+import '../service/firestore.dart';
+
+@immutable
+class GetStatusFirestore extends RequestHandler<Body> {
+  const GetStatusFirestore({
+    required super.config,
+    @visibleForTesting this.datastoreProvider = DatastoreService.defaultProvider,
+    @visibleForTesting this.buildStatusProvider = BuildStatusService.defaultProvider,
+  });
+
+  final DatastoreServiceProvider datastoreProvider;
+  final BuildStatusServiceProvider buildStatusProvider;
+
+  static const String kLastCommitKeyParam = 'lastCommitKey';
+  static const String kBranchParam = 'branch';
+  static const String kRepoParam = 'repo';
+
+  @override
+  Future<Body> get() async {
+    final String? encodedLastCommitKey = request!.uri.queryParameters[kLastCommitKeyParam];
+    final String repoName = request!.uri.queryParameters[kRepoParam] ?? Config.flutterSlug.name;
+    final RepositorySlug slug = RepositorySlug('flutter', repoName);
+    final String branch = request!.uri.queryParameters[kBranchParam] ?? Config.defaultBranch(slug);
+    final DatastoreService datastore = datastoreProvider(config.db);
+    final FirestoreService firestoreService = await config.createFirestoreService();
+    final BuildStatusService buildStatusService = buildStatusProvider(datastore, firestoreService);
+    final KeyHelper keyHelper = config.keyHelper;
+    final int commitNumber = config.commitNumber;
+    final int lastCommitTimestamp = await _obtainTimestamp(encodedLastCommitKey, keyHelper, datastore);
+
+    final List<SerializableCommitTasksStatus> statuses = await buildStatusService
+        .retrieveCommitStatusFirestore(
+          limit: commitNumber,
+          timestamp: lastCommitTimestamp,
+          branch: branch,
+          slug: slug,
+        )
+        .map<SerializableCommitTasksStatus>(
+          (CommitTasksStatus status) => SerializableCommitTasksStatus(status),
+        )
+        .toList();
+
+    return Body.forJson(<String, dynamic>{
+      'Statuses': statuses,
+    });
+  }
+
+  Future<int> _obtainTimestamp(String? encodedLastCommitKey, KeyHelper keyHelper, DatastoreService datastore) async {
+    /// [lastCommitTimestamp] is initially set as the current time, which allows query return
+    /// latest commit list. If [owerKeyParam] is not empty, [lastCommitTimestamp] will be set
+    /// as the timestamp of that [commit], and the query will return lastest commit
+    /// list whose timestamp is smaller than [lastCommitTimestamp].
+    int lastCommitTimestamp = DateTime.now().millisecondsSinceEpoch;
+
+    if (encodedLastCommitKey != null) {
+      final Key<String> ownerKey = keyHelper.decode(encodedLastCommitKey) as Key<String>;
+      final Commit commit = await datastore.db.lookupValue<Commit>(
+        ownerKey,
+        orElse: () => throw NotFoundException('Failed to find commit with key $ownerKey'),
+      );
+
+      lastCommitTimestamp = commit.timestamp!;
+    }
+    return lastCommitTimestamp;
+  }
+}
+
+class SerializableCommitTasksStatus {
```

Comment:
Should we consider moving this to its own class file? Will it potentially have uses outside of this class and do you think it makes sense to move it closer to the CommitTaskStatus class?

Reply:
Good point. Created a new model file `commit_tasks_status.dart`, and moved logics related to `CommitTasksStatus` there.